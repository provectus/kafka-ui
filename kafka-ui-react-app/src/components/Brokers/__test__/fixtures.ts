import { BrokersLogdirs } from 'generated-sources';

export const brokersPayload = [
  { id: 1, host: 'b-1.test.kafka.amazonaws.com', port: 9092 },
  { id: 2, host: 'b-2.test.kafka.amazonaws.com', port: 9092 },
];

export const clusterStatsPayload = {
  brokerCount: 2,
  activeControllers: 1,
  onlinePartitionCount: 138,
  offlinePartitionCount: 0,
  inSyncReplicasCount: 239,
  outOfSyncReplicasCount: 0,
  underReplicatedPartitionCount: 0,
  diskUsage: [
    { brokerId: 0, segmentSize: 334567, segmentCount: 245 },
    { brokerId: 1, segmentSize: 12345678, segmentCount: 121 },
  ],
  version: '2.2.1',
};

const partition = {
  broker: 2,
  offsetLag: 0,
  partition: 2,
  size: 0,
};
const topics = {
  name: '_confluent-ksql-devquery_CTAS_NUMBER_OF_TESTS_59-Aggregate-Aggregate-Materialize-changelog',
  partitions: [partition],
};

export const brokerLogDirsPayload: BrokersLogdirs[] = [
  {
    error: 'NONE',
    name: '/opt/kafka/data-0/logs',
    topics: [
      {
        ...topics,
        partitions: [partition, partition, partition],
      },
      topics,
      {
        ...topics,
        partitions: [],
      },
    ],
  },
];
