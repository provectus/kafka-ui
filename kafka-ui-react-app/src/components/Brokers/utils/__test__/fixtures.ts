import { BrokerMetrics } from 'generated-sources';

export const brokerMetricsPayload: BrokerMetrics = {
  segmentSize: 23,
  segmentCount: 23,
  metrics: [
    {
      name: 'TotalFetchRequestsPerSec',
      labels: {
        canonicalName:
          'kafka.server:name=TotalFetchRequestsPerSec,topic=_connect_status,type=BrokerTopicMetrics',
      },
      value: 10,
    },
    {
      name: 'ZooKeeperRequestLatencyMs',
      value: 11,
    },
    {
      name: 'RequestHandlerAvgIdlePercent',
    },
  ],
};
export const transformedBrokerMetricsPayload =
  '{"segmentSize":23,"segmentCount":23,"metrics":[{"name":"TotalFetchRequestsPerSec","labels":{"canonicalName":"kafka.server:name=TotalFetchRequestsPerSec,topic=_connect_status,type=BrokerTopicMetrics"},"value":10},{"name":"ZooKeeperRequestLatencyMs","value":11},{"name":"RequestHandlerAvgIdlePercent"}]}';
