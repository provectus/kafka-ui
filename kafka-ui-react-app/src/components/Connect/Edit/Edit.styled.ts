import styled from 'styled-components';

export const ConnectEditWrapperStyled = styled.div`
  margin: 16px;

  & form > *:last-child {
    margin-top: 16px;
  }
`;

export const ConnectEditWarningMessageStyled = styled.div`
  height: 48px;
  display: flex;
  align-items: center;
  background-color: ${({ theme }) => theme.connectEditWarning};
  border-radius: 8px;
  padding: 8px;
  margin-bottom: 16px;
`;
