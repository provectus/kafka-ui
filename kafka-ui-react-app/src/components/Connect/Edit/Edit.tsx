import React from 'react';
import { useHistory, useParams } from 'react-router-dom';
import { Controller, useForm } from 'react-hook-form';
import { ErrorMessage } from '@hookform/error-message';
import { yupResolver } from '@hookform/resolvers/yup';
import { Connector } from 'generated-sources';
import {
  ClusterName,
  ConnectName,
  ConnectorConfig,
  ConnectorName,
} from 'redux/interfaces';
import { clusterConnectConnectorConfigPath } from 'lib/paths';
import yup from 'lib/yupExtended';
import JSONEditor from 'components/common/JSONEditor/JSONEditor';
import PageLoader from 'components/common/PageLoader/PageLoader';
import { Button } from 'components/common/Button/Button';

import {
  ConnectEditWarningMessageStyled,
  ConnectEditWrapperStyled,
} from './Edit.styled';

const validationSchema = yup.object().shape({
  config: yup.string().required().isJsonObject(),
});

interface RouterParams {
  clusterName: ClusterName;
  connectName: ConnectName;
  connectorName: ConnectorName;
}

interface FormValues {
  config: string;
}

export interface EditProps {
  fetchConfig(
    clusterName: ClusterName,
    connectName: ConnectName,
    connectorName: ConnectorName
  ): Promise<void>;
  isConfigFetching: boolean;
  config: ConnectorConfig | null;
  updateConfig(
    clusterName: ClusterName,
    connectName: ConnectName,
    connectorName: ConnectorName,
    connectorConfig: ConnectorConfig
  ): Promise<Connector | undefined>;
}

const Edit: React.FC<EditProps> = ({
  fetchConfig,
  isConfigFetching,
  config,
  updateConfig,
}) => {
  const { clusterName, connectName, connectorName } = useParams<RouterParams>();
  const history = useHistory();
  const {
    handleSubmit,
    control,
    formState: { isDirty, isSubmitting, isValid, errors },
    setValue,
  } = useForm<FormValues>({
    mode: 'onTouched',
    resolver: yupResolver(validationSchema),
    defaultValues: {
      config: JSON.stringify(config, null, '\t'),
    },
  });

  React.useEffect(() => {
    fetchConfig(clusterName, connectName, connectorName);
  }, [fetchConfig, clusterName, connectName, connectorName]);

  React.useEffect(() => {
    if (config) {
      setValue('config', JSON.stringify(config, null, '\t'));
    }
  }, [config, setValue]);

  const onSubmit = React.useCallback(
    async (values: FormValues) => {
      const connector = await updateConfig(
        clusterName,
        connectName,
        connectorName,
        JSON.parse(values.config.trim())
      );
      if (connector) {
        history.push(
          clusterConnectConnectorConfigPath(
            clusterName,
            connectName,
            connectorName
          )
        );
      }
    },
    [updateConfig, clusterName, connectName, connectorName]
  );

  if (isConfigFetching) return <PageLoader />;

  const hasCredentials = JSON.stringify(config, null, '\t').includes(
    '"******"'
  );
  return (
    <ConnectEditWrapperStyled>
      {hasCredentials && (
        <ConnectEditWarningMessageStyled>
          Please replace ****** with the real credential values to avoid
          accidentally breaking your connector config!
        </ConnectEditWarningMessageStyled>
      )}
      <form onSubmit={handleSubmit(onSubmit)} aria-label="Edit connect form">
        <div>
          <Controller
            control={control}
            name="config"
            render={({ field }) => (
              <JSONEditor {...field} readOnly={isSubmitting} />
            )}
          />
        </div>
        <div>
          <ErrorMessage errors={errors} name="config" />
        </div>
        <Button
          buttonSize="M"
          buttonType="primary"
          type="submit"
          disabled={!isValid || isSubmitting || !isDirty}
        >
          Submit
        </Button>
      </form>
    </ConnectEditWrapperStyled>
  );
};

export default Edit;
