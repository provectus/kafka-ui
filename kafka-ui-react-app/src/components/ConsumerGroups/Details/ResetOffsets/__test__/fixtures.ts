export const expectedOutputs = {
  EARLIEST: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: undefined,
        partition: 0,
      },
    ],
    resetType: 'EARLIEST',
    topic: '__amazon_msk_canary',
  },
  LATEST: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: undefined,
        partition: 0,
      },
    ],
    resetType: 'LATEST',
    topic: '__amazon_msk_canary',
  },
  OFFSET: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: '10',
        partition: 0,
      },
    ],
    resetType: 'OFFSET',
    topic: '__amazon_msk_canary',
  },
};
