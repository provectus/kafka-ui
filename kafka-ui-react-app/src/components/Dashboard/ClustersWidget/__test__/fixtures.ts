import { Cluster, ServerStatus } from 'generated-sources';

export const onlineCluster: Cluster = {
  name: 'secondLocal',
  defaultCluster: false,
  status: ServerStatus.Online,
  brokerCount: 1,
  onlinePartitionCount: 6,
  topicCount: 3,
  bytesInPerSec: 0.000030618196853764715,
  bytesOutPerSec: 5.737800890036267075817,
};

export const offlineCluster: Cluster = {
  name: 'local',
  defaultCluster: true,
  status: ServerStatus.Offline,
  brokerCount: 1,
  onlinePartitionCount: 2,
  topicCount: 2,
  bytesInPerSec: 8000.0000067376808542600021,
  bytesOutPerSec: 0.8153063567297119490871,
};

export const clusters: Cluster[] = [onlineCluster, offlineCluster];
