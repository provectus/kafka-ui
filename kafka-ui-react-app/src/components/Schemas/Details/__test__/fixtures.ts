import { SchemaSubject, SchemaType } from 'generated-sources';
import {
  schemaVersion1,
  schemaVersion2,
  schemaVersionWithNonAsciiChars,
} from 'redux/reducers/schemas/__test__/fixtures';

export const versionPayload = [
  schemaVersion1,
  schemaVersion2,
  schemaVersionWithNonAsciiChars,
];
export const versionEmptyPayload = [];

export const jsonSchema: SchemaSubject = {
  subject: 'test',
  version: '15',
  id: 1,
  schema:
    '{"type":"record","name":"MyRecord1","namespace":"com.mycompany","fields":[{"name":"id","type":"long"}]}',
  compatibilityLevel: 'BACKWARD',
  schemaType: SchemaType.JSON,
};

export const protoSchema: SchemaSubject = {
  subject: 'test_proto',
  version: '1',
  id: 2,
  schema:
    'syntax = "proto3";\npackage com.indeed;\n\nmessage MyRecord {\n  int32 id = 1;\n  string name = 2;\n}\n',
  compatibilityLevel: 'BACKWARD',
  schemaType: SchemaType.PROTOBUF,
};
