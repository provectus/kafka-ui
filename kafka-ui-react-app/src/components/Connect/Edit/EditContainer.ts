import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { RootState } from 'redux/interfaces';
import { fetchConnectorConfig, updateConnectorConfig } from 'redux/actions';
import {
  getConnectorConfig,
  getIsConnectorConfigFetching,
} from 'redux/reducers/connect/selectors';

import Edit from './Edit';

const mapStateToProps = (state: RootState) => ({
  isConfigFetching: getIsConnectorConfigFetching(state),
  config: getConnectorConfig(state),
});

const mapDispatchToProps = {
  fetchConfig: fetchConnectorConfig,
  updateConfig: updateConnectorConfig,
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Edit));
