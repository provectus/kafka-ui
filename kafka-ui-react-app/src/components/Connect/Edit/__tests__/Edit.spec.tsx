import React from 'react';
import { render } from 'lib/testHelpers';
import {
  clusterConnectConnectorConfigPath,
  clusterConnectConnectorEditPath,
} from 'lib/paths';
import Edit, { EditProps } from 'components/Connect/Edit/Edit';
import { connector } from 'redux/reducers/connect/__test__/fixtures';
import { Route } from 'react-router-dom';
import { waitFor } from '@testing-library/dom';
import { act, fireEvent, screen } from '@testing-library/react';

jest.mock('components/common/PageLoader/PageLoader', () => 'mock-PageLoader');

jest.mock('components/common/Editor/Editor', () => 'mock-Editor');

const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

describe('Edit', () => {
  const pathname = clusterConnectConnectorEditPath();
  const clusterName = 'my-cluster';
  const connectName = 'my-connect';
  const connectorName = 'my-connector';

  const renderComponent = (props: Partial<EditProps> = {}) =>
    render(
      <Route path={pathname}>
        <Edit
          fetchConfig={jest.fn()}
          isConfigFetching={false}
          config={connector.config}
          updateConfig={jest.fn()}
          {...props}
        />
      </Route>,
      {
        pathname: clusterConnectConnectorEditPath(
          clusterName,
          connectName,
          connectorName
        ),
      }
    );

  it('fetches config on mount', async () => {
    const fetchConfig = jest.fn();
    await waitFor(() => renderComponent({ fetchConfig }));
    expect(fetchConfig).toHaveBeenCalledTimes(1);
    expect(fetchConfig).toHaveBeenCalledWith({
      clusterName,
      connectName,
      connectorName,
    });
  });

  it('calls updateConfig on form submit', async () => {
    const updateConfig = jest.fn();
    await waitFor(() => renderComponent({ updateConfig }));
    fireEvent.submit(screen.getByRole('form'));
    await waitFor(() => expect(updateConfig).toHaveBeenCalledTimes(1));
    expect(updateConfig).toHaveBeenCalledWith({
      clusterName,
      connectName,
      connectorName,
      connectorConfig: connector.config,
    });
  });

  it('redirects to connector config view on successful submit', async () => {
    const updateConfig = jest.fn().mockResolvedValueOnce(connector);
    await waitFor(() => renderComponent({ updateConfig }));
    fireEvent.submit(screen.getByRole('form'));

    await waitFor(() => expect(mockHistoryPush).toHaveBeenCalledTimes(1));
    expect(mockHistoryPush).toHaveBeenCalledWith(
      clusterConnectConnectorConfigPath(clusterName, connectName, connectorName)
    );
  });

  it('does not redirect to connector config view on unsuccessful submit', async () => {
    const updateConfig = jest.fn().mockResolvedValueOnce(undefined);
    await waitFor(() => renderComponent({ updateConfig }));
    await act(() => {
      fireEvent.submit(screen.getByRole('form'));
    });
    expect(mockHistoryPush).not.toHaveBeenCalled();
  });
});
