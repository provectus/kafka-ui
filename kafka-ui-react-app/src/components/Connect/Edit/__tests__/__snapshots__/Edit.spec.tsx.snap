// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit view matches snapshot 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0px 12px;
  border: none;
  border-radius: 4px;
  white-space: nowrap;
  background: #4F4FFF;
  color: #FFFFFF;
  font-size: 14px;
  font-weight: 500;
  height: 32px;
}

.c1:hover:enabled {
  background: #1717CF;
  color: #FFFFFF;
  cursor: pointer;
}

.c1:active:enabled {
  background: #1414B8;
  color: #FFFFFF;
}

.c1:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c1 a {
  color: white;
}

.c1 i {
  margin-right: 7px;
}

.c0 {
  margin: 16px;
}

.c0 form > *:last-child {
  margin-top: 16px;
}

<div
  className="c0"
>
  <form
    onSubmit={[Function]}
  >
    <div>
      <mock-JSONEditor
        name="config"
        onBlur={[Function]}
        onChange={[Function]}
        readOnly={false}
        value="{
	\\"connector.class\\": \\"FileStreamSource\\",
	\\"tasks.max\\": \\"10\\",
	\\"topic\\": \\"test-topic\\",
	\\"file\\": \\"/some/file\\"
}"
      />
    </div>
    <div />
    <button
      className="c1"
      disabled={true}
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Edit view matches snapshot when config has credentials 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0px 12px;
  border: none;
  border-radius: 4px;
  white-space: nowrap;
  background: #4F4FFF;
  color: #FFFFFF;
  font-size: 14px;
  font-weight: 500;
  height: 32px;
}

.c2:hover:enabled {
  background: #1717CF;
  color: #FFFFFF;
  cursor: pointer;
}

.c2:active:enabled {
  background: #1414B8;
  color: #FFFFFF;
}

.c2:disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

.c2 a {
  color: white;
}

.c2 i {
  margin-right: 7px;
}

.c0 {
  margin: 16px;
}

.c0 form > *:last-child {
  margin-top: 16px;
}

.c1 {
  height: 48px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #FFEECC;
  border-radius: 8px;
  padding: 8px;
  margin-bottom: 16px;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    Please replace ****** with the real credential values to avoid accidentally breaking your connector config!
  </div>
  <form
    onSubmit={[Function]}
  >
    <div>
      <mock-JSONEditor
        name="config"
        onBlur={[Function]}
        onChange={[Function]}
        readOnly={false}
        value="{
	\\"connector.class\\": \\"FileStreamSource\\",
	\\"tasks.max\\": \\"10\\",
	\\"topic\\": \\"test-topic\\",
	\\"file\\": \\"/some/file\\",
	\\"password\\": \\"******\\"
}"
      />
    </div>
    <div />
    <button
      className="c2"
      disabled={true}
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`Edit view matches snapshot when fetching config 1`] = `<mock-PageLoader />`;
